db/hints/manager.cc

```diff
@@ -56,6 +56,13 @@ const std::string manager::FILENAME_PREFIX("HintsLog" + commitlog::descriptor::S
 const std::chrono::seconds manager::hint_file_write_timeout = std::chrono::seconds(2);
 const std::chrono::seconds manager::hints_flush_period = std::chrono::seconds(10);
 
+const std::chrono::seconds manager::end_point_hints_manager::sender::concurrency_limiter::update_interval = std::chrono::seconds(1);
+const std::chrono::seconds manager::end_point_hints_manager::sender::concurrency_limiter::update_delay_after_timeout = std::chrono::seconds(10);
+
+thread_local std::uniform_int_distribution<uint64_t> manager::end_point_hints_manager::sender::concurrency_limiter::_next_delay_distribution{1, 120};
```

Comment:
`std::uniform_int_distribution` is a very thin wrapper class which just keeps the min and max value (ref: https://github.com/gcc-mirror/gcc/blob/8f323c712ea76cc4506b03895e9b991e4e4b2baf/libstdc%2B%2B-v3/include/bits/uniform_int_dist.h#L76-L95) . Thus, consider if it's not better to just place in on the stack where it's actually used, because then it has a high chance of getting optimized out as a couple of constants. As a thread_local, you risk that the compiler is going to wrap the access to this variable with tls guarding code, which would be detrimental to performance - since you're basically guarding a pair of 2 const integers.

Reply:
I'll convert it to a non-static field. Technically speaking, an implementation of a distribution may keep some internal state because `operator()` is not marked as const. Other distributions do keep state and creating temporaries might be costly or give incorrect results. If we decide to change the distribution in the future, it will be easier if we keep a non-static field.

A link to an article with some more context: https://quuxplusone.github.io/blog/2019/10/22/psa-stateful-distributions/

As a side note, the `static thread_local std::uniform_int_distribution` pattern already occurs several times in the code. Moreover, this variable will usually be accessed at most once per second, for each endpoint manager which actively sends hints - so I'm not sure if it makes much of a difference.